## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [186]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [187]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [152]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [188]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [189]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)

#df.groupby('category').size().nlargest(3) #This is another way to get the same answer

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [190]:
# For the category with the most entries, which subcategory had the most entries?
top_category_name = df['category'].value_counts().index[0]
print(f'Top category: {top_category_name}')
print()
print('Top subcategory:')
df[df['category'] == top_category_name]['subcategory'].value_counts().head(1)


Top category: consumables

Top subcategory:


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [191]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [192]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head().index.tolist()
print(top_5_clients)

[33615, 66037, 46820, 38378, 24741]


In [193]:
# How many total units (the qty column) did the client with the most entries order order?
client_counts = df.groupby('client_id')['order_id'].count()

top_client_id = client_counts.sort_values(ascending=False).index[0]
print(f"Top Client ID with most entries: {top_client_id}")

total_units_ordered = df[df['client_id'] == top_client_id]['qty'].sum()

print(f"Total Units Ordered: {total_units_ordered}")

Top Client ID with most entries: 33615
Total Units Ordered: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [194]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [195]:
# Function to calculate shipping weight
def shipping_weight(unit_weight,qty):
    return unit_weight * qty

# Function to calculate shipping cost
def shipping_cost(shipping_weight):
    return 7 if shipping_weight > 50 else 10

# Function to calculate shipping price
def shipping_price(unit_weight, qty):
    weight = shipping_weight(unit_weight, qty)
    cost_per_pound = shipping_cost(weight)
    return weight * cost_per_pound

# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df['shipping_price'] = df.apply(lambda row: shipping_price(row['unit_weight'], row['qty']), axis=1)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  \
0     1096.80     762.71         7.50  105            1      115164.00   
1       24.95      15.09         1.49   21            0         523.95   
2       13.52       7.86         1.68   39            6         527.28   
3       36.42      24.85         1.23   29            3        1056.18   
4      195.10     108.17        46.43   20            1        3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [196]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = df.apply(lambda row: round((row['line_subtotal'] + row['shipping_price']) * 1.0925, 2), axis = 1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  line_subtotal  shipping_price  \
0     762.71         7.50  105            1      115164.00         5512.50   
1      15.09         1.49   21            0         523.95          312.90   
2       7.86         1.68   39            6         527.28          458.64   
3      24.85         1.23   29            3        1056.18          356.70   
4     108.17        46.43   20            1        3902.00         6500.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 21 columns]

In [197]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df.apply(lambda row: (row['unit_cost'] * row['qty']) + row['shipping_price'], axis = 1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  shipping_price  total_price  \
0         7.50  105            1      115164.00         5512.50    131839.08   
1         1.49   21            0         523.95          312.90       914.26   
2         1.68   39            6         527.28          458.64      1077.12   
3         1.23   29            3        1056.18          356.70      1543.57   
4        46.43   20            1        3902.00         6500.20     11364.40   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 22 columns]

In [198]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df.apply(lambda row: row['total_price'] - row['line_cost'], axis=1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  shipping_price  total_price  line_cost  \
0            1      115164.00         5512.50    131839.08   85597.05   
1            0         523.95          312.90       914.26     629.79   
2            6         527.28          458.64      1077.12     765.18   
3            3        1056.18          356.70      1543.57    1077.35   
4            1        3902.00         6500.20     11364.40    8663.60   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  
3       466.22  
4      2700.80  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [199]:
# # Check your work using the totals above
#APPROACH 1
# order_1 = 2742071
# order_2 = 2173913
# order_3 = 6128929

# order_1_filtered = df[df['order_id'] == order_1]
# order_2_filtered = df[df['order_id'] == order_2]
# order_3_filtered = df[df['order_id'] == order_3]

# total_price_for_order_1 = order_1_filtered['total_price'].sum()
# total_price_for_order_2 = order_2_filtered['total_price'].sum()
# total_price_for_order_3 = order_3_filtered['total_price'].sum()

# print(f'Order {order_1}: Total: ${total_price_for_order_1:,.2f}')
# print(f'Order {order_2}: Total: ${total_price_for_order_2:,.2f}')
# print(f'Order {order_3}: Total: ${total_price_for_order_3:,.2f}')

# APPROACH 2
total_prices = {} #Initialize an empty dictionary to hold the total price for each order to sum the price

order_id_list_to_sum_up = [2742071, 2173913, 6128929] #list of orders to sum up total price

for order_id in order_id_list_to_sum_up: #Loop through each order ID
    filtered_order = df[df['order_id'] == order_id] #Filter the data for the current order ID
    total_price = filtered_order['total_price'].sum() #Sum up the total_price for the current order ID
    total_prices[order_id] = total_price #Store the total_price in total prices list

for order_id in order_id_list_to_sum_up:
    print(f'Order {order_id}: Total: ${total_prices[order_id]:,.2f}') #Print the total prices for the specified order IDs




Order 2742071: Total: $152,811.89
Order 2173913: Total: $162,388.71
Order 6128929: Total: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [200]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

# APPROACH 1
# top_clients = df['client_id'].value_counts().head().index.tolist() # Get the top clients by quantity

# client_1, client_2, client_3, client_4, client_5 = top_clients # Assign clien_id to variables

# client_1_df = df[df['client_id'] == client_1] # store all records for each client in a dataframe variable
# client_2_df = df[df['client_id'] == client_2]
# client_3_df = df[df['client_id'] == client_3]
# client_4_df = df[df['client_id'] == client_4]
# client_5_df = df[df['client_id'] == client_5]

# client_1_total_spend = client_1_df['total_price'].sum() # Get the total spend for each client
# client_2_total_spend = client_2_df['total_price'].sum()
# client_3_total_spend = client_3_df['total_price'].sum()
# client_4_total_spend = client_4_df['total_price'].sum()
# client_5_total_spend = client_5_df['total_price'].sum()

# print("Total Spend of 'Top 5 Client By Quantity'")
# print(f'{client_1}: ${client_1_total_spend:,.2f}')
# print(f'{client_2}: ${client_2_total_spend:,.2f}')
# print(f'{client_3}: ${client_3_total_spend:,.2f}')
# print(f'{client_4}: ${client_4_total_spend:,.2f}')
# print(f'{client_5}: ${client_5_total_spend:,.2f}')

# APPROACH 2
top_5_clients_total_spend = {} # Empty list to hold the total spend for each of the top 5 clients

for client_id in top_5_clients: #Loop through each order ID
    filtered_lines_client = df[df['client_id'] == client_id] # Filter the data for the client ID
    client_total_spend = filtered_lines_client['total_price'].sum() #Sum up the total_price for the current client ID
    top_5_clients_total_spend[client_id] = client_total_spend #Store the total_price in the top_5_clients_total_spend list

for client_id in top_5_clients:
    print(f'{client_id}: ${top_5_clients_total_spend[client_id]:,.2f}')




33615: $8,377,308.53
66037: $10,259,514.80
46820: $9,743,794.33
38378: $12,906,550.87
24741: $82,268,892.04


In [201]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5_clients_summary = [] # Empty list to hold summary data

for client_id in top_5_clients:
    filtered_lines_client = df[df['client_id'] == client_id] # Filter the data for the client ID
    client_total_units_purchased = filtered_lines_client['qty'].sum() # Sum up the total units purchased
    client_total_shipping_price = filtered_lines_client['shipping_price'].sum() # Sum up the total shipping price
    client_total_revenue = filtered_lines_client['total_price'].sum() # Sum up the total revenue
    client_total_linecost = filtered_lines_client['line_cost'].sum() # Sum up the line cost
    client_total_profit = filtered_lines_client['line_profit'].sum() # Sum up the total profit
    top_5_clients_summary.append({
        'client_id': client_id,
        'qty': client_total_units_purchased,
        'shipping_price': client_total_shipping_price,
        'line_price': client_total_revenue,
        'line_cost': client_total_linecost,
        'line_profit': client_total_profit
    })

top_5_clients_summary_df = pd.DataFrame(top_5_clients_summary)
top_5_clients_summary_df.head()

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.53   6175313.91   2201994.62
1      66037   43018      1395151.85  10259514.80   7004482.98   3255031.82
2      46820   75768      1601448.84   9743794.33   7007191.64   2736602.69
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.04  45688899.71  36579992.33

In [203]:
# Format the data and rename the columns to names suitable for presentation.
top_5_clients_summary_df = top_5_clients_summary_df.rename(columns={
    "client_id": "Client ID", 
    "qty": "Units", 
    "shipping_price": "Shipping Price", 
    "line_price": "Total Revenue", 
    "line_cost": "Total Cost", 
    "line_profit": "Total Profit"
})

# Define the money columns. 
money_columns = ["Shipping Price", "Total Revenue", "Total Cost", "Total Profit"]

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    converted_value = value/1000000
    return converted_value

# Apply the currency_format_millions function to only the money columns. 
for column_name in money_columns:
    top_5_clients_summary_df[column_name] = top_5_clients_summary_df[column_name].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
renamed_df_columns = top_5_clients_summary_df.rename(columns={
    "Shipping Price": "Shipping (millions)",
    "Total Revenue": "Total Revenue (millions)",
    "Total Cost": "Total Cost (millions)",
    "Total Profit": "Total Profit (millions)"
})

renamed_df_columns


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             0.000002                  0.000008   
1      66037   43018             0.000001                  0.000010   
2      46820   75768             0.000002                  0.000010   
3      38378   73667             0.000003                  0.000013   
4      24741  239862             0.000005                  0.000082   

   Total Cost (millions)  Total Profit (millions)  
0               0.000006                 0.000002  
1               0.000007                 0.000003  
2               0.000007                 0.000003  
3               0.000010                 0.000003  
4               0.000046                 0.000037

In [204]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_renamed_df_columns = renamed_df_columns.sort_values("Total Profit (millions)", ascending=False)
print(sorted_renamed_df_columns)

   Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             0.000005                  0.000082   
3      38378   73667             0.000003                  0.000013   
1      66037   43018             0.000001                  0.000010   
2      46820   75768             0.000002                  0.000010   
0      33615   64313             0.000002                  0.000008   

   Total Cost (millions)  Total Profit (millions)  
4               0.000046                 0.000037  
3               0.000010                 0.000003  
1               0.000007                 0.000003  
2               0.000007                 0.000003  
0               0.000006                 0.000002  
